src/pyramid/static.py

```diff
@@ -94,46 +143,126 @@ def __call__(self, context, request):
-            filepath = normcase(normpath(join(self.norm_docroot, path)))
-            if isdir(filepath):
+            resource_path = normcase(normpath(join(self.norm_docroot, path)))
+            if isdir(resource_path):
                 if not request.path_url.endswith('/'):
-                    self.add_slash_redirect(request)
-                filepath = join(filepath, self.index)
-            if not exists(filepath):
-                raise HTTPNotFound(request.url)
+                    raise self.add_slash_redirect(request)
+                resource_path = join(resource_path, self.index)
 
-        content_type, content_encoding = _guess_type(filepath)
-        return FileResponse(
-            filepath,
-            request,
-            self.cache_max_age,
-            content_type,
-            content_encoding=None,
-        )
+        return resource_path
+
+    def find_resource_path(self, name):
+        """
+        Return the absolute path to the resource or ``None`` if it doesn't
+        exist.
+
+        """
+        if self.package_name:
+            if resource_exists(self.package_name, name):
+                return resource_filename(self.package_name, name)
+
+        elif exists(name):
+            return name
+
+    def get_possible_files(self, resource_name):
+        """ Return a sorted list of ``(size, encoding, path)`` entries."""
+        result = self.filemap.get(resource_name)
+        if result is not None:
+            return result
+
+        # XXX we could put a lock around this work but worst case scenario a
+        # couple requests scan the disk for files at the same time and then
+        # the cache is set going forward so do not bother
+        result = []
+
+        # add the identity
+        path = self.find_resource_path(resource_name)
+        if path:
+            result.append((path, None))
+
+        # add each file we find for the supported encodings
+        # we don't mind adding multiple files for the same encoding if there
+        # are copies with different extensions because we sort by size so the
+        # smallest is always found first and the rest ignored
+        for encoding, extensions in self.content_encodings.items():
+            for ext in extensions:
+                encoded_name = resource_name + ext
+                path = self.find_resource_path(encoded_name)
+                if path:
+                    result.append((path, encoding))
+
+        # sort the files by size, smallest first
+        result.sort(key=lambda x: getsize(x[0]))
+
+        # only cache the results if reload is disabled
+        if not self.reload:
+            self.filemap[resource_name] = result
+        return result
+
+    def find_best_match(self, request, files):
+        """ Return ``(path | None, encoding)``."""
+        # if the client did not specify encodings then assume only the
+        # identity is acceptable
+        if not request.accept_encoding:
```

Comment:
short circuit `if len(files) < 2` here and return the identity path as well - no sense in doing the rest of the accept-header negotiation if we only have one possible result